tests/rbd/test_performance_immutable_cache.py

```diff
@@ -0,0 +1,249 @@
+
+    Raises:
+        Exception: If the performance with cache is not better than without cache.
+    """
+
+    def calculate_total_time(time_data):
+        """
+        Calculate the total time for write and read operations.
+
+        Args:
+            time_data (str): Time data for an operation.
+
+        Returns:
+            float: Total time in seconds.
+        """
+        real_line = [line for line in time_data.split("\n") if line.startswith("real")][
+            0
+        ]
+        real_minutes, real_seconds = map(
+            float, real_line.split("\t")[1][:-1].split("m")
+        )
+        real_time = real_minutes * 60 + real_seconds
+
+        return real_time
+
+    total_time_with_cache = calculate_total_time(
+        time_with_cache[1]
+    ) + calculate_total_time(time_with_cache[3])
+
+    total_time_without_cache = calculate_total_time(
+        time_without_cache[1]
+    ) + calculate_total_time(time_without_cache[3])
+
+    log.info(f"Time taken for IO operation with cache: {total_time_with_cache}")
+    log.info(f"Time taken for IO operation without cache: {total_time_without_cache}")
+
+    # Compare performance
+    if total_time_without_cache > total_time_with_cache:
+        log.info("Time taken without cache is greater than with cache")
+        log.info("Performance is better with cache enabled")
+    else:
+        log.error("Time taken to without cache is lesser than with cache")
+        raise Exception("Performance validation with/without cache failed")
+
+
+def perform_clone_io(client, rbd_obj, pool, image, **kw):
+    """
+    Perform I/O operations on cloned images and return the time taken.
+
+    Args:
+        client (object): Client node object.
+        rbd (object): RBD object.
+        config (dict): Configuration dictionary.
+        pool (str): Pool name.
+        image (str): Image name.
+
+    Returns:
+        str: Time taken for the I/O operation.
+    """
+    config = kw.get("config")
+    rbd = rbd_obj.get("rbd")
+
+    # Write Data to the Parent Image
+    fio_args = {
+        "client_node": client,
+        "pool_name": pool,
+        "image_name": image,
+        "io_type": "write",
+        "size": config["fio"]["size"],
+    }
+
+    run_fio(**fio_args)
+
+    # Create snapshot, protect, and clone for the image
+    snap = "image_snap" + random_string(len=5)
+    clone = "image_clone" + random_string(len=5)
+    snap_spec = f"{pool}/{image}@{snap}"
+    clone_spec = f"{pool}/{clone}"
+
+    create_snap_and_clone(rbd, snap_spec, clone_spec)
+
+    # After image clone performing read operation from parent and clone
+
+    fio_args = {
+        "client_node": client,
+        "pool_name": pool,
+        "image_name": image,
+        "io_type": "read",
+        "size": config["fio"]["size"],
+        "get_time_taken": True,
+    }
+
+    config["time_taken"] = run_fio(**fio_args)
+    time = config.get("time_taken", "")
```

Comment:
And also one major change, when Ilya mentioned the below steps.
`3) map, read the entire clone, unmap
4) map, read the entire clone again, unmap`

It meant, you create an image, fill with io, create a clone, read io from clone once (time_1), read io from the same clone again (time_2) by mapping and unmapping compare time_1 and time_2.

When cache is enabled time_2 should be less than time_1 because when you read second time it is reading it from stored cache.
When cache is not enabled time_2 and time_1 will be same, because every read reads from the actual image from scratch.

Your logic is very different from this. Could you please correct this?

Reply:
thanks for explaining the logic i bit confused here now implemented the code accordingly.